README.md
# PlexPVRChannel.bundle

CHANGELOG
11/04/16 - 1.0.0 - Initial Checkin


Contents/Tests/ChannelTests.py
from plex_test_case import PlexTestCase
from Framework.api.objectkit import ObjectContainer

class MyChannelTest(PlexTestCase):

    def test_main_menu(self):
        # Load './MyChannelTest/MainMenuObjectContainer.xml' as expected XML generated by channel.
        expected = self.get_file_contents('MainMenuObjectContainer.xml')
        # Fix HTTP response from the './MyChannelTest/Index.html' file contents.
        self.networking.http_response_body = self.get_file_contents('Index.html')
        # Make a request to channel code (which makes a HTTP request mocked above).
        status, headers, body = self.request('/video/mychannel')
        # Check the return values against expected.
        self.assertEquals(200, status)
        self.assertEquals('application/xml', headers['Content-Type'])
        self.assertEquals(expected, body)

Contents/Tests/plex_test_case/mock.py
import imp
import os

mock_path = os.path.join(os.path.dirname(__file__), 'mock')
f, filename, desc = imp.find_module('mock', [mock_path])
mock = imp.load_module('mock', f, filename, desc)


Contents/Tests/plex_test_case/plex.py
import inspect
import logging
import os
import sys
import urllib2
from Framework import components, core
from StringIO import StringIO


GLOBAL_DEFAULT_TIMEOUT = components.networking.GLOBAL_DEFAULT_TIMEOUT

class BasicRequest(object):

    def __init__(self, uri, headers={}, method='GET', body=''):
        self.uri = uri
        self.headers = headers
        self.method = method
        self.body = body


class FrameworkCore(core.FrameworkCore):
    """
        Framework core class extension to modify components in testing environment.
    """
    def _setup_storage(self):
        # Don't configure output into real log file.
        logging.basicConfig()
        logging.setLoggerClass(core.FrameworkLogger)
        logger = logging.getLogger()
        logger.handlers[0].addFilter(core.LogFilter())
        self.log = logging.getLogger(self.identifier)
        self._start_component(components.Storage)
        self._values = {}
        self._values_file_path = os.path.join(self.plugin_support_path, 'Data', self.identifier, 'StoredValues')

    def get_server_info(self):
        # Don't do actual request to Plex server and return dummy values.
        self.attributes['machineIdentifier'] = 'Unit testing'
        self.attributes['serverVersion'] = None

    def start(self):
        # Same implementation as in parent class, just without error handling, to allow us to see what went wrong.
        if self.init_code:
            self.sandbox.execute(self.init_code)
        self.sandbox.call_named_function(core.START_FUNCTION_NAME)

    def log_exception(self, fmt, *args):
        error = sys.exc_info()
        if error:
            # Catch a special case in the original `Runtime.handle_request`. If error is allowed to raise here, then
            # the local variable `body` will not be set and the code fail on `UnboundLocalError` in `finally` later.
            (frame, _, _, function_name, _, _) = inspect.getouterframes(inspect.currentframe())[1]
            if function_name == 'handle_request' and frame.f_locals['self'].__class__ == Runtime:
                self.runtime.request_error = error
                return
            # In general case, just raise the error.
            raise error[0], error[1], error[2]

    def _start_component(self, component):
        # Replace runtime and networking components with our implementations.
        if component == components.Runtime:
            component = Runtime
        elif component == components.Networking:
            component = Networking
        super(FrameworkCore, self)._start_component(component)


class PreferenceManager(object):

    def __init__(self, values={}):
        self._dict = values

    def get(self):
        return self._dict


class ResponseHeaders(object):

    def __init__(self, headers):
        self.dict = {k: v.lower() for k, v in headers.iteritems()}

    def get(self, header, default=None):
        lowercase = header.lower()
        return self.dict[lowercase] if lowercase in self.dict else default


class ResponseBody(StringIO):

    def __init__(self, string):
        StringIO.__init__(self, string)
        self.recv = None

    @property
    def _sock(self):
        return self


class HTTPHandler(urllib2.HTTPHandler):

    def __init__(self, networking):
        self._networking = networking

    def get_response(self, req):
        if self._networking.http_response_body is not None:
            response = urllib2.addinfourl(ResponseBody(self._networking.http_response_body),
                                          ResponseHeaders(self._networking.http_response_headers),
                                          req.get_full_url())
            response.code = 200
            response.msg = "OK"
            return response
        raise RuntimeError('Unhandled HTTP request: %s' % req.get_full_url())

    def http_open(self, req):
        return self.get_response(req)


class Networking(components.Networking):
    """
        Networking class extension to avoid sending actual requests in tests.
    """
    def _init(self):
        super(Networking, self)._init()
        self.http_response_body = None
        self.http_response_headers = {}

    def build_opener(self, cookie_jar=None):
        return urllib2.build_opener(HTTPHandler(self))


class Runtime(components.Runtime):
    """
        Runtime class extension to keep track of the framework's threads and locks.
    """
    def _init(self):
        self._threads = []
        super(Runtime, self)._init()

    def create_thread(self, f, log=True, sandbox=None, globalize=True, *args, **kwargs):
        # Store the created threads so that their status can be checked after tests have run.
        thread = super(Runtime, self).create_thread(f, log, sandbox, globalize, *args, **kwargs)
        self._threads.append(thread)
        return thread

    def check_threads(self):
        # Check for alive non-daemon threads and throw error if found.
        # Ignored threads are just slow but they don't run in a loop so they'll finish eventually.
        ignored = ['_setup_shared_code_sandbox']
        for thread in self._threads:
            if thread.is_alive() and not thread.daemon and thread.name not in ignored:
                raise RuntimeError('Alive non-daemon thread %s found, this will prevent shutdown.' % str(thread))

    def handle_request(self, request):
        # In case the parent method encounters an error reported using `_report_error` method, we'll raise it here.
        self.request_error = None
        status, headers, body = super(Runtime, self).handle_request(request)
        if self.request_error is not None:
            raise self.request_error[0], self.request_error[1], self.request_error[2]
        return status, headers, body

    def _report_error(self, request, controller, kwargs, e):
        return

    def get_resource_hashes(self):
        return


Contents/Code/Keyboard.py
#
#
#
# MOST IMPORTANT NOTE: BEFORE WRITING A CHANNEL, THERE MUST ALREADY BE A URL SERVICE FOR THE VIDEOS ON THE WEBSITE
# YOU WANT TO CREATE A CHANNEL FOR OR YOU WILL HAVE TO WRITE A URL SERVICE BEFORE YOU CAN WRITE THE CHANNEL. TO
# SEE IF A URL SERVICE ALREADY EXISTS, CHECK THE SERVICES BUNDLE IN THE PLEX PLUGIN FOLDER
#
# IMPORTANT NOTE: PYTHON IS VERY SENSITIVE TO PROPER INDENTIONS.  IF YOUR CHANNEL HAS IMPROPER INDENTIONS IT WILL
# NOT BE RECOGNIZED BY PLEX. I RUN THE PROGRAM THROUGH A CHECK MODULE ON A LOCAL VERSION OF PYTHON I HAVE LOADED
# PRIOR TO ACCESSING IT THROUGH PLEX TO MAKE SURE THERE ARE NO INDENTION ERRORS.
#
# Keyboard.py by Dave Spriet <davespriet@gmail.com>
# 
# Change History
#    v1.0.0 Modified version of DumbTools for Plex v1.1 
#

KEYS = list('abcdefghijklmnopqrstuvwxyz1234567890-=;[]\\\',./')
SHIFT_KEYS = list('ABCDEFGHIJKLMNOPQRSTUVWXYZ!@#$%^&*()_+:{}|\"<>?')

TITLE = 'Plex Request Channel'
PREFIX = '/video/plexpvr'


DUMB_KEYBOARD_CLIENTS = ['Plex for iOS', 'Plex Media Player', 'Plex Home Theater', 'OpenPHT', 'Plex for Roku', 'iOS', 'Roku', 'tvOS' 'Konvergo',
                         'Plex for Apple TV', 'Plex for Xbox 360', 'Plex for Xbox One', 'Xbox One', 'Mystery 4', 'Plex for Vizio', 'Plex TV']

NO_MESSAGE_CONTAINER_CLIENTS = ['Plex for iOS', 'tvOS', 'Plex for Apple TV', 'Plex for Xbox One', 'iOS', 'Mystery 4', 'Samsung', 'Plex for Samsung']

MESSAGE_OVERLAY_CLIENTS = ['Android', 'Roku', 'Konvergo', 'Plex Web']


@route(PREFIX + "/dumbtools/keyboard")
def Keyboard(query=None, callback=None, parent_call=None, shift=False, secure='False', locked='locked', title="Search", message=None):
    if 'DumbKeyboard-History' not in Dict:
        Dict['DumbKeyboard-History'] = []

    if not query:
        Log.Debug("Keyboard created.")
    if secure == 'True' and query is not None:
        string = ''.join(['*' for i in range(len(query[:-1]))]) + query[-1]
    else:
        string = query if query else ""
    if Client.Platform in MESSAGE_CONTAINER_CLIENTS or Client.Product in MESSAGE_CONTAINER_CLIENTS:
        oc = ObjectContainer(title1=title, title2=string, message=message)
    else:
        oc = ObjectContainer(title1=title, title2=string)
    # Submit
    # Log.Debug("Create Submit key")
    oc.add(DirectoryObject(key=Callback(callback, query=query, locked=locked), title=u'%s: %s' % (L('Submit'), string)))
    # Search History
    if Dict['DumbKeyboard-History']:
        # Log.Debug("Create History")
        oc.add(
            DirectoryObject(key=Callback(History, query=query, callback=callback, locked=locked, secure=secure), title=u'%s' % L('Search History')))
    # Space
    # Log.Debug("Create Space Key")
    oc.add(DirectoryObject(key=Callback(Keyboard, query=query + " " if query else " ", callback=callback,parent_call=parent_call, locked=locked, secure=secure, title=title),
                           title=L('Space')))
    # Backspace (not really needed since you can just hit back)
    if query is not None:
        # Log.Debug("Create backspace key")
        oc.add(DirectoryObject(
            key=Callback(Keyboard, query=query[:-1], callback=callback, parent_call=parent_call, locked=locked, secure=secure, title=title),
            title=L('Backspace')))
    # Shift
    # Log.Debug("Create Shift Key")
    oc.add(
        DirectoryObject(
            key=Callback(Keyboard, query=query, callback=callback, parent_call=parent_call, locked=locked, secure=secure, shift=True, title=title),
            title=L('Shift')))
    # Keys
    # Log.Debug("Generating keys")
    if parent_call:
        oc.add(DirectoryObject(key=parent_call, title=L("Cancel")))
    for key in KEYS if not shift else SHIFT_KEYS:
        oc.add(
            DirectoryObject(key=Callback(Keyboard, query=query + key if query else key, callback=callback, parent_call=parent_call, locked=locked, secure=secure, title=title),
                            title=u'%s' % key))
    # Log.Debug("Return Object Container")
    return oc


@route(PREFIX + "/dumbtools/history")
def History(query=None, callback=None, parent_call=None, locked='locked', secure='False'):
    oc = ObjectContainer()
    oc.add(DirectoryObject(key=Callback(Keyboard, query=query, callback=callback, parent_call=parent_call, locked=locked, secure=secure),
                           title=L("Return to Keyboard")))
    if Dict['DumbKeyboard-History']:
        oc.add(DirectoryObject(key=Callback(ClearHistory, callback=callback, parent_call=parent_call, locked=locked, secure=secure),
                               title=u'%s' % L('Clear History')))
    for item in Dict['DumbKeyboard-History']:
        oc.add(DirectoryObject(key=Callback(callback, query=item, locked=locked),
                               title=u'%s' % item))
    return oc


@route(PREFIX + "/dumbtools/clearhistory")
def ClearHistory(query="", callback=None, parent_call=None, locked='locked', secure='False'):
    Dict['DumbKeyboard-History'] = []
    Dict.Save()
    return Keyboard(query=query, callback=callback, parent_call=parent_call, locked=locked, secure=secure)


@route(PREFIX + "/dumbtools/addhistory")
def AddHistory(query):
    if query not in Dict['DumbKeyboard-History']:
        Dict['DumbKeyboard-History'].append(query)
        Dict.Save()

# @route(PREFIX + "/dumbtools/submit")
# def Submit(query, callback=callback, locked=locked):
#     self.AddHistory(query)
#     kwargs = {'query': query}
#     kwargs.update(self.callback_args)
#     return Callback(**kwargs)


Contents/Code/DumbTools.py
# DumbTools for Plex v1.1 by Dave Spriet <davespriet@gmail.com>
# Slightly modified by ngovil21 for Plex Request Channel

import urllib2

#NO_MESSAGE_CONTAINER_CLIENTS is now deprecated, using known clients that work with message overlays instead
NO_MESSAGE_CONTAINER_CLIENTS = ['Plex for iOS', 'tvOS', 'Plex for Apple TV', 'Plex for Xbox One', 'iOS', 'Mystery 4', 'Samsung',
                                'Plex for Samsung']
MESSAGE_OVERLAY_CLIENTS = ['Android', 'Roku', 'Konvergo', 'Plex Web']


class DumbKeyboard:
    CLIENTS = ['Plex for iOS', 'Plex Media Player', 'Plex for Roku', 'iOS', 'Roku', 'tvOS' 'Konvergo',
               'Plex for Apple TV', 'Plex for Xbox 360', 'Plex for Xbox One', 'Xbox One', 'Mystery 4', 'Plex for Vizio', 'Plex TV']

    KEYS = list('abcdefghijklmnopqrstuvwxyz1234567890-=;[]\\\',./')
    SHIFT_KEYS = list('ABCDEFGHIJKLMNOPQRSTUVWXYZ!@#$%^&*()_+:{}|\"<>?')

    def __init__(self, prefix, oc, callback, parent_call=None, dktitle=None, dkthumb=None,
                 dkplaceholder=None, dksecure=False, message=None, **kwargs):
        cb_hash = hash(str(callback) + str(kwargs))
        Log.Debug("Generated callback hash: " + str(cb_hash))
        Route.Connect(prefix + '/dumbkeyboard/%s/keyboard' % cb_hash, self.Keyboard)
        Route.Connect(prefix + '/dumbkeyboard/%s/submit' % cb_hash, self.Submit)
        Route.Connect(prefix + '/dumbkeyboard/%s/history' % cb_hash, self.History)
        Route.Connect(prefix + '/dumbkeyboard/%s/history/clear' % cb_hash, self.ClearHistory)
        Route.Connect(prefix + '/dumbkeyboard/%s/history/add/{query}' % cb_hash, self.AddHistory)
        Log.Debug("Generated Keyboard Routes")
        # Log.Debug(str(Callback(self.Keyboard)))
        # Add our directory item
        oc.add(DirectoryObject(key=Callback(self.Keyboard, query=dkplaceholder, message=message),
                               title=str(dktitle) if dktitle else u'%s' % 'DumbKeyboard Search', thumb=dkthumb))

        if 'DumbKeyboard-History' not in Dict:
            Dict['DumbKeyboard-History'] = []

        self.Callback = callback
        self.callback_args = kwargs
        self.secure = dksecure
        self.parent_call = parent_call

    def Keyboard(self, query=None, shift=False, message=None):
        # Log.Debug("Keyboard created.")
        if self.secure and query is not None:
            string = ''.join(['*' for i in range(len(query[:-1]))]) + query[-1]
        else:
            string = query if query else ""

        oc = ObjectContainer()
        if message and (Client.Product in MESSAGE_OVERLAY_CLIENTS or Client.Platform in MESSAGE_OVERLAY_CLIENTS):
            oc.message = message
        # Submit
        # Log.Debug("Create Submit key")
        oc.add(DirectoryObject(key=Callback(self.Submit, query=query),
                               title=u'%s: %s' % ('Submit', string)))
        #                      title=u'%s: %s' % ('Submit', string.replace(' ', '_')))), -why replace with underscores?
        # Search History
        if Dict['DumbKeyboard-History']:
            # Log.Debug("Create History")
            oc.add(DirectoryObject(key=Callback(self.History),
                                   title=u'%s' % 'Search History'))
        # Space
        # Log.Debug("Create Space Key")
        oc.add(DirectoryObject(key=Callback(self.Keyboard,
                                            query=query + " " if query else " "),
                               title='Space'))
        # Backspace (not really needed since you can just hit back)
        if query is not None:
            # Log.Debug("Create backspace key")
            oc.add(DirectoryObject(key=Callback(self.Keyboard, query=query[:-1]),
                                   title='Backspace'))
        # Shift
        # Log.Debug("Create Shift Key")
        oc.add(DirectoryObject(key=Callback(self.Keyboard, query=query, shift=True),
                               title='Shift'))
        # Cancel - return to parent
        if self.parent_call:
            oc.add(DirectoryObject(key=self.parent_call, title="Cancel"))
        # Keys
        # Log.Debug("Generating keys")
        for key in self.KEYS if not shift else self.SHIFT_KEYS:
            oc.add(DirectoryObject(key=Callback(self.Keyboard,
                                                query=query + key if query else key),
                                   title=u'%s' % key))
        # Log.Debug("Return Object Container")
        return oc

    def History(self):
        oc = ObjectContainer()
        if Dict['DumbKeyboard-History']:
            oc.add(DirectoryObject(key=Callback(self.ClearHistory),
                                   title=u'%s' % 'Clear History'))
        for item in Dict['DumbKeyboard-History']:
            oc.add(DirectoryObject(key=Callback(self.Submit, query=item),
                                   title=u'%s' % item))
        return oc

    def ClearHistory(self):
        Dict['DumbKeyboard-History'] = []
        Dict.Save()
        return self.History()

    def AddHistory(self, query):
        if query not in Dict['DumbKeyboard-History']:
            Dict['DumbKeyboard-History'].append(query)
            Dict.Save()

    def Submit(self, query):
        self.AddHistory(query)
        kwargs = {'query': query}
        kwargs.update(self.callback_args)
        return self.Callback(**kwargs)


class DumbPrefs:
    clients = ['Plex for iOS', 'Plex Media Player', 'Plex Home Theater',
               'OpenPHT', 'Plex for Roku']

    def __init__(self, prefix, oc, title=None, thumb=None):
        self.host = 'http://127.0.0.1:32400'
        try:
            self.CheckAuth()
        except Exception as e:
            Log.Error('DumbPrefs: this user cant access prefs: %s' % str(e))
            return

        Route.Connect(prefix + '/dumbprefs/list', self.ListPrefs)
        Route.Connect(prefix + '/dumbprefs/listenum', self.ListEnum)
        Route.Connect(prefix + '/dumbprefs/set', self.Set)
        Route.Connect(prefix + '/dumbprefs/settext', self.SetText)
        oc.add(DirectoryObject(key=Callback(self.ListPrefs),
                               title=title if title else 'Preferences',
                               thumb=thumb))
        self.prefix = prefix
        self.GetPrefs()

    def GetHeaders(self):
        headers = Request.Headers
        headers['Connection'] = 'close'
        return headers

    def CheckAuth(self):
        """ Only the main users token is accepted at /myplex/account """
        headers = {'X-Plex-Token': Request.Headers.get('X-Plex-Token', '')}
        req = urllib2.Request("%s/myplex/account" % self.host, headers=headers)
        res = urllib2.urlopen(req)

    def GetPrefs(self):
        data = HTTP.Request("%s/:/plugins/%s/prefs" % (self.host, Plugin.Identifier),
                            headers=self.GetHeaders())
        prefs = XML.ElementFromString(data).xpath('/MediaContainer/Setting')

        self.prefs = [{'id': pref.xpath("@id")[0],
                       'type': pref.xpath("@type")[0],
                       'label': pref.xpath("@label")[0],
                       'default': pref.xpath("@default")[0],
                       'secure': True if pref.xpath("@secure")[0] == "true" else False,
                       'values': pref.xpath("@values")[0].split("|") \
                           if pref.xpath("@values") else None
                       } for pref in prefs]

    def Set(self, key, value):
        HTTP.Request("%s/:/plugins/%s/prefs/set?%s=%s" % (self.host,
                                                          Plugin.Identifier,
                                                          key, value),
                     headers=self.GetHeaders(),
                     immediate=True)
        return ObjectContainer()

    def ListPrefs(self):
        oc = ObjectContainer(no_cache=True)
        for pref in self.prefs:
            do = DirectoryObject()
            value = Prefs[pref['id']] if not pref['secure'] else \
                ''.join(['*' for i in range(len(Prefs[pref['id']]))])
            title = u'%s: %s = %s' % (L(pref['label']), pref['type'], L(value))
            if pref['type'] == 'enum':
                do.key = Callback(self.ListEnum, id=pref['id'])
            elif pref['type'] == 'bool':
                do.key = Callback(self.Set, key=pref['id'],
                                  value=str(not Prefs[pref['id']]).lower())
            elif pref['type'] == 'text':
                if Client.Product in DumbKeyboard.clients:
                    DumbKeyboard(self.prefix, oc, self.SetText,
                                 id=pref['id'],
                                 dktitle=title,
                                 dkplaceholder=Prefs[pref['id']],
                                 dksecure=pref['secure'])
                else:
                    oc.add(InputDirectoryObject(key=Callback(self.SetText, id=pref['id']),
                                                title=title))
                continue
            else:
                do.key = Callback(self.ListPrefs)
            do.title = title
            oc.add(do)
        return oc

    def ListEnum(self, id):
        oc = ObjectContainer()
        for pref in self.prefs:
            if pref['id'] == id:
                for i, option in enumerate(pref['values']):
                    oc.add(DirectoryObject(key=Callback(self.Set, key=id, value=i),
                                           title=u'%s' % option))
        return oc

    def SetText(self, query, id):
        return self.Set(key=id, value=query)


Contents/Code/__init__.py
#
# MOST IMPORTANT NOTE: BEFORE WRITING A CHANNEL, THERE MUST ALREADY BE A URL SERVICE FOR THE VIDEOS ON THE WEBSITE
# YOU WANT TO CREATE A CHANNEL FOR OR YOU WILL HAVE TO WRITE A URL SERVICE BEFORE YOU CAN WRITE THE CHANNEL. TO
# SEE IF A URL SERVICE ALREADY EXISTS, CHECK THE SERVICES BUNDLE IN THE PLEX PLUGIN FOLDER

# IMPORTANT NOTE: PYTHON IS VERY SENSITIVE TO PROPER INDENTIONS.  IF YOUR CHANNEL HAS IMPROPER INDENTIONS IT WILL
# NOT BE RECOGNIZED BY PLEX. I RUN THE PROGRAM THROUGH A CHECK MODULE ON A LOCAL VERSION OF PYTHON I HAVE LOADED
# PRIOR TO ACCESSING IT THROUGH PLEX TO MAKE SURE THERE ARE NO INDENTION ERRORS.

# You will need to decide how you want to set up your channel. If you want to have just one page that list all 
# the videos or if you want to break these videos down into subsections for different types of videos, individual shows, season, etc
# It is easiest to determine this system based on the structure of the website you are accessing. 

# You can hard code these choice in or pull the data from a web page or JSON data file and put it in a for loop to 
# automate the process. I created a basic example in the form of functions below to show the most common methods of 
# parsing data from different types of websites. When you want to produce results to the screen and have subpage come up # when they click on those results, you usually will use a
# DirectoryObject and include the name of the next function that will create that subpage called in the key.
# The key callback section sends your data to the next function that you will use to produce your next subpage.  Usually
# you will pass the value of the url onto your next function, but there are many attributes that can be sent.  It is good 
# to pass the title as well so it shows up at the top of the screen. Refer to the Framework Documentation to see the full
# list

# You will need a good working knowledge of xpath the parse the data properly.  Good sources for information related to 
# xpath are:
# 'http://devblog.plexapp.com/2012/11/14/xpath-for-channels-the-good-the-bad-and-the-fugly/'
# 'http://forums.plexapp.com/index.php/topic/49086-xpath-coding/'

# Here is a good article about working with Chrome Development Tools: 
# 'http://devblog.plexapp.com/2012/09/27/using-chromes-built-in-debugger-for-channel-development/'

# And here are a few pages that give you some pointers ON figuring out the basics of creating a channel
# 'http://devblog.plexapp.com/2011/11/16/a-beginners-guide-to-v2-1/'
# 'http://forums.plexapp.com/index.php/topic/28084-plex-plugin-development-walkthrough/'

# The title of your channel should be unique and as explanatory as possible.  The preifx is used for the channel
# store and shows you where the channel is executed in the log files
#
# __init__.py by Dave Spriet <davespriet@gmail.com>
# 
# Change History
#    v1.0.0 Initial Release
#

from Keyboard import Keyboard, DUMB_KEYBOARD_CLIENTS, MESSAGE_OVERLAY_CLIENTS
from DumbTools import DumbKeyboard, MESSAGE_OVERLAY_CLIENTS

import re

TITLE = 'Plex Request Channel'
PREFIX = '/video/plexpvr'

# The images below are the default graphic for your channel and should be saved or located in you Resources folder
# The art and icon should be a certain size for channel submission. The graphics should be good quality and not be blurry
# or pixelated. Icons must be 512x512 PNG files and be named, icon-default.png. The art must be 1280x720 JPG files and be
# named, art-default.jpg. The art shows up in the background of the PMC Client, so you want to make sure image you choose 
# is not too busy or distracting.  I tested out a few in PMC to figure out which one looked best.

ART = 'art-default.jpg'
ICON = 'plexpvr.png'

from Session import VERSION

CHANGELOG_URL = "https://raw.githubusercontent.com/ufctester/PlexPVRChannel.bundle/master/CHANGELOG"

### URL Constants for TheMovieDataBase ############################################################
TMDB_API_KEY = "096c49df1d0974ee573f0295acb9e3ce"
TMDB_API_URL = "http://api.themoviedb.org/3/"
TMDB_IMAGE_BASE_URL = "http://image.tmdb.org/t/p/"
POSTER_SIZE = "w500/"
BACKDROP_SIZE = "original/"
###################################################################################################

### URL Constants for OpenMovieDataBase 
OMDB_API_URL = "http://www.omdbapi.com/"
###################################################################################################

### URL Constants for TheTVDB 
TVDB_API_KEY = "B93EF22D769A70CB"
TVDB_API_URL = "http://thetvdb.com/api/"
TVDB_BANNER_URL = "http://thetvdb.com/banners/"
###################################################################################################

### Notification Constants 
PUSHBULLET_API_URL = "https://api.pushbullet.com/v2/"
PUSHOVER_API_URL = "https://api.pushover.net/1/messages.json"
PUSHOVER_API_KEY = "ajMtuYCg8KmRQCNZK2ggqaqiBw2UHi"
###################################################################################################

TV_SHOW_OBJECT_FIX_CLIENTS = ["Android", "Plex for Android"]

from LocalePatch import SetAvailableLanguages

LANGUAGES = ['en']

###################################################################################################
#   Start Code
#
# This (optional) function is initially called by the PMS framework to initialize the plug-in. This includes setting up
# the Plug-in static instance along with the displayed artwork. These setting below are pretty standard
# You first set up the containers and default for all possible objects.  You will probably mostly use Directory Objects
# and Videos Objects. But many of the other objects give you added entry fields you may want to use and set default thumb
# and art for. For a full list of objects and their parameters, refer to the Framework Documentation.
########################################################
def Start():

# There are  few commands you may see appear in this section that are no longer needed.  Below is an explanation of them
# provided from a very helpful channel designer who was nice enough to explain their purpose to me:
#    Plugin.AddViewGroup("List", viewMode="List", mediaType="items")
# This a left-over from when plugins had more control over how the contents were displayed. ViewGroups were defined to 
# tell the client how to display the contents of the directory. Now, most (if not all) clients have a fairly rigid model
# for what types of content get display in which way. Generally, it is best to remove it from a plugins when since it
# gets ignored anyways. 
#
#    HTTP.CacheTime = CACHE_1HOUR
# This setting a global cache time for all HTTP requests made by the plugin. This over-rides the framework's default 
# cache period which,
# I don't remember off the top of my head. It is entirely optional, but if you're going to use it, the idea is to pick a 
# cache-time that is reasonable. Ie. store data for a long enough time that you can realistically reduce the load on
# external servers as well as speed up the load-time for HTTP-requests, but not so long that changes/additions are not
# caught in a reasonable time frame. IMO, unless you specifically want/need a specific cache-length, I would remove that
# line and allow the framework to manage the cache with its default settings.
#
#   HTTP.Headers['User-Agent'] = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.8; rv:18.0) Gecko/20100101 Firefox/18.0'
# This assigning a specific User-Agent header for all HTTP requests made by the plugin. Generally speaking, each time a
# plugin is started, a user-agent is randomly selected from a list to be used for all HTTP requests. In some cases, a
# plugin will perform better using a specific user-agent instead of a randomly assigned one. For example, some websites
# return different data for Safari on an iPad, then what they return for Chrome or Firefox on a PC. Again, if you don't
# have a specific need to set a specific user-agent, I would remove that code from your channel.

# You set up the default attributes for all you object containers and objects.  You will probably mostly use Directory
# Objects and Videos Objects but many of the other objects give you added entry fields you may want to use.  For a full 
# list of objects and their parameters, refer to the Framework Documentation.

# Important Note: (R stands for Resources folder) to tell the channel where these images are located.

    ObjectContainer.title1 = TITLE
    ObjectContainer.art = R(ART)

    DirectoryObject.thumb = R(ICON)
    DirectoryObject.art = R(ART)
    EpisodeObject.thumb = R(ICON)
    EpisodeObject.art = R(ART)
    VideoClipObject.thumb = R(ICON)
    VideoClipObject.art = R(ART)

    Plugin.AddViewGroup("Details", viewMode="InfoList", mediaType="items")
    Plugin.AddViewGroup("List", viewMode="List", mediaType="items")

    Log.Debug("Channel Version: " + VERSION)

    SetAvailableLanguages(LANGUAGES)

    # DICT[] 
    # DICT[] IS PART OF THE PLEX FRAMEWORK THAT ALLOWS YOU TO SAVE DATA IN A GLOBAL VARIABLE THAT IS RETAINED WHEN YOU EXIT THE PLUGIN
    # SO YOU CAN PULL IT UP IN MULTIPLE FUNCTIONS WITHOUT PASSING THE VARIABLES FROM FUNCTION TO FUNCTION. AND IT CAN BE ACCESSED AND USED
    # OVER MULTIPLE SESSIONS
    ## Initialize Dictionary
    if 'tv' not in Dict:
        Dict['tv'] = {}
    if 'movie' not in Dict:
        Dict['movie'] = {}
    if 'music' not in Dict:
        Dict['music'] = {}
    if 'register' not in Dict:
        Dict['register'] = {}
        Dict['register_reset'] = Datetime.TimestampFromDatetime(Datetime.Now())
    if 'blocked' not in Dict:
        Dict['blocked'] = []
    if 'sonarr_users' not in Dict:
        Dict['sonarr_users'] = []
    if 'sickbeard_users' not in Dict:
        Dict['sickbeard_users'] = []
    if 'debug' not in Dict:
        Dict['debug'] = False
    if 'DumbKeyboard-History' not in Dict:
        Dict['DumbKeyboard-History'] = []
    Dict.Save()


def ValidatePrefs():
    return


from Session import Session


###################################################################################################
# This tells Plex how to list you in the available channels and what type of channels this is 
@handler(PREFIX, TITLE, art=ART, thumb=ICON)
@route(PREFIX + '/main')

def MainMenu():
    Log.Info("=================== __init__::MainMenu ==================")
    sesh = Session(session_id=Hash.MD5(str(Datetime.Now())))
    return sesh.SMainMenu()

"""
List of Client.Product and Client.Platform

Client.Product 	        Description
--------------          ------------------------------------------------------------------
Plex for Android 	    Android phone
Plex for iOS 	        Apple phone
Plex Home Theater 	    This is for Plex Home Theater
Plex Media Player 	    This is for new Plex Media Player
Plex Web 	            Plex Web client, from web browser
Plex for Firefox TV 	Plex Firefox TV, (source), this one is a guess
Plex for Roku 	        Roku
OpenPHT 	            OpenPHT
Plex Chromecast 	    Chromecast
NotifyPlex 	            NZBGet
HTPC Manager 	        Windows Server (Windows-2012Server-6.2.9200). Not sure if correct.
Plex for Xbox One 	    Xbox One
Plex for Xbox 360       Xbox 360
Plex for Samsung        Samsung

Client.Platform 	    Description
---------------         -------------------------------------------------------------------
Android 	            Android phone
iOS 	                Apple phone
Safari 	                This is a guess from looking at a Service URL code.
Chrome              	Plex Web client on Chrome internet browser
Plex Home Theater 	    This is for Plex Home Theater
Konvergo 	            Plex Media Player, (running on a Raspberry PI 2 B)
tvOS 	                New AppleTV
MacOSX 	                MacOSX
Linux 	                Linux
Windows 	            Windows
LGTV 	                LGTV
Roku 	                Roku
Chromecast          	Chromecast
NZBGet 	                NZBGet
Xbox One            	Xbox One
Xbox 360                Xbox 360
Samsung                 Samsung

"""
